File: oss/src/main/java/com/truenine/component/oss/amazon/S3PolicyCreator.java
package com.truenine.component.oss.amazon;

import com.google.gson.Gson;


public class S3PolicyCreator {
  private S3PolicyCreator() {
  }


  public static S3BuilderChain builder() {
    return new S3BuilderChain();
  }

  public static S3BuilderChain publicBucket(String bucketName) {
    S3PrincipalArgs p = new S3PrincipalArgs();
    p.getAws().add("*");
    return new S3BuilderChain()
      .addStatement(S3StatementBuilder.builder()
        .principal(p)
        .addAction(S3Policies.Bucket.LIST_MUL_UPLOADS)
        .addAction(S3Policies.Bucket.LIST)
        .addAction(S3Policies.Bucket.GET_LOCATION)
        .addResource(bucketName)
      )
      .addStatement(S3StatementBuilder.builder()
        .principal(p)
        .addAction(S3Policies.Obj.GET)
        .addAction(S3Policies.Obj.LIST_MUL_UPLOAD_PARTS)
        .addAction(S3Policies.Obj.PUT)
        .addAction(S3Policies.Obj.ABORT_MUL_UPLOAD)
        .addAction(S3Policies.Obj.DEL)
        .addResource(bucketName + "/*")
      );
  }

  public static S3BuilderChain publicBucketAndReadOnly(String bucketName) {
    S3PrincipalArgs p = new S3PrincipalArgs();
    p.getAws().add("*");
    return new S3BuilderChain()
      .addStatement(S3StatementBuilder.builder()
        .principal(p)
        .addAction(S3Policies.Bucket.LIST)
        .addAction(S3Policies.Bucket.GET_LOCATION)
        .addResource(bucketName)
      )
      .addStatement(S3StatementBuilder.builder()
        .principal(p)
        .addAction(S3Policies.Obj.GET)
        .addResource(bucketName + "/*")
      );
  }

  public static class S3BuilderChain {

    private final S3Args RULE;

    private final Gson GSON = new Gson();


    private S3BuilderChain() {
      this.RULE = new S3Args();
      RULE.setVersion("2012-10-17");
    }

    public S3BuilderChain addStatement(S3StatementBuilder builder) {
      RULE.getStatement().add(builder.statement());
      return this;
    }

    public S3BuilderChain version(String version) {
      RULE.setVersion(version);
      return this;
    }


    public String json() {
      return GSON.toJson(RULE, S3Args.class);
    }
  }

  public static class S3StatementBuilder {

    private final S3StatementArgs S = new S3StatementArgs();

    private S3StatementBuilder() {
    }

    public static S3StatementBuilder builder() {
      return new S3StatementBuilder();
    }

    private S3StatementArgs statement() {
      return S;
    }

    public S3StatementBuilder principal(S3PrincipalArgs principal) {
      S.setPrincipal(principal);
      return this;
    }

    public S3StatementBuilder addResource(String resource) {
      S.getResource().add("arn:aws:s3:::" + resource);
      return this;
    }

    public S3StatementBuilder addAction(String action) {
      S.getAction().add("s3:" + action);
      return this;
    }

    public S3StatementBuilder effect(String effect) {
      S.setEffect(effect);
      return this;
    }


    public S3StatementBuilder build() {
      return this;
    }
  }
}

File: oss/common/src/main/kotlin/net/yan100/compose/oss/common/S3PolicyCreator.kt
package net.yan100.compose.oss.common

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper

object S3PolicyCreator {
  fun builder(): S3BuilderChain {
    return S3BuilderChain()
  }

  fun publicBucket(bucketName: String): S3BuilderChain {
    val p = S3PrincipalArgs()
    p.aws.add("*")
    return S3BuilderChain()
      .addStatement(
        S3StatementBuilder.builder()
          .effect(S3Policies.Effect.ALLOW)
          .principal(p)
          .addAction(S3Policies.Bucket.LIST_MUL_UPLOADS)
          .addAction(S3Policies.Bucket.LIST)
          .addAction(S3Policies.Bucket.GET_LOCATION)
          .addResource(bucketName)
      )
      .addStatement(
        S3StatementBuilder.builder()
          .principal(p)
          .effect(S3Policies.Effect.ALLOW)
          .addAction(S3Policies.Object.GET)
          .addAction(S3Policies.Object.LIST_MUL_UPLOAD_PARTS)
          .addAction(S3Policies.Object.PUT)
          .addAction(S3Policies.Object.ABORT_MUL_UPLOAD)
          .addAction(S3Policies.Object.DEL)
          .addResource("$bucketName/*")
      )
  }

  fun privateBucket(bucketName: String): S3BuilderChain {
    val p = S3PrincipalArgs()
    p.aws.add("*")
    return S3BuilderChain()
      .addStatement(
        S3StatementBuilder.builder()
          .principal(p)
          .effect(S3Policies.Effect.ALLOW)
          .addAction(S3Policies.Bucket.LIST)
          .addAction(S3Policies.Bucket.GET_LOCATION)
          .addResource(bucketName)
      )
      .addStatement(
        S3StatementBuilder.builder()
          .principal(p)
          .effect(S3Policies.Effect.ALLOW)
          .addAction(S3Policies.Object.GET)
          .addResource("$bucketName/*")
      )
  }

  fun publicReadonlyBucket(bucketName: String): S3BuilderChain {
    val p = S3PrincipalArgs()
    p.aws.add("*")
    return S3BuilderChain()
      .addStatement(
        S3StatementBuilder.builder()
          .principal(p)
          .effect(S3Policies.Effect.ALLOW)
          .addAction(S3Policies.Bucket.LIST)
          .addAction(S3Policies.Bucket.GET_LOCATION)
          .addResource(bucketName)
      )
      .addStatement(
        S3StatementBuilder.builder()
          .principal(p)
          .effect(S3Policies.Effect.ALLOW)
          .addAction(S3Policies.Object.GET)
          .addResource("$bucketName/*")
      )
  }

  class S3BuilderChain {
    private val RULE = S3Args()
    private val objectMapper = ObjectMapper()

    init {
      RULE.version = "2012-10-17"
    }

    fun addStatement(builder: S3StatementBuilder): S3BuilderChain {
      RULE.statement.add(builder.statement())
      return this
    }

    fun version(version: String?): S3BuilderChain {
      RULE.version = version
      return this
    }

    fun json(): String {
      try {
        return objectMapper.writeValueAsString(RULE)
      } catch (e: JsonProcessingException) {
        throw RuntimeException(e)
      }
    }
  }

  class S3StatementBuilder private constructor() {
    private val S = S3StatementArgs()

    fun statement(): S3StatementArgs {
      return S
    }

    fun principal(principal: S3PrincipalArgs?): S3StatementBuilder {
      S.principal = principal
      return this
    }

    fun addResource(resource: String): S3StatementBuilder {
      S.resource.add("arn:aws:s3:::$resource")
      return this
    }

    fun addAction(action: String): S3StatementBuilder {
      S.action.add("s3:$action")
      return this
    }

    fun effect(effect: String?): S3StatementBuilder {
      S.effect = effect
      return this
    }

    fun build(): S3StatementBuilder {
      return this
    }

    companion object {
      fun builder(): S3StatementBuilder {
        return S3StatementBuilder()
      }
    }
  }
}
